crates/pallet-domains/src/lib.rs

```diff
@@ -116,6 +116,16 @@ mod pallet {
         /// Delay before a domain runtime is upgraded.
         type DomainRuntimeUpgradeDelay: Get<Self::BlockNumber>;
 
+        /// The block tree pruning depth, its value should >= `BlockHashCount` because we
```

Comment:
There is a mistake, the assumption should be `BlockTreePruningDepth::get() <= BlockHashCount::get()`, and I found that this assumption is not enough to guarantee consensus block hash exists while verifying receipt because the domain block is not mapping to consensus block one by one, a domain can produce domain block 1 on consensus block 1 and produce domain block 2 at consensus block 1000...

Thus we may still need to manually store the consensus hash, but by utilizing `SuccessfulBundles` we can detect if there will be a domain block driving from the consensus block and only need to store the consensus block hash in this case. In the above case we only need to store consensus block hash of block 1 and block 1000 instead of all te hash from block 1 to block 1000 like v1 does.

Reply:
That does not alleviate the problem honestly, it's expected that there are bundles in most consensus blocks for a domain, we run into the same rabbit hole as before if tracking the consensus block hash in the state, which was an open problem in v1 if you remember. We may have to introduce a host function to check it on the client side.

Reply:
> The gap will be filled up eventually given the assumption that there is at least one honest operator

It won't because ER is submitted with a bundle and there is only one ER within a bundle. In the best case, the ER extends the head receipt to reduce the gap by one but the bundle will result in a domain block thus increase the gap by one, so the gap remains.

My point is that the hash is used to verify upcoming ER in the first place, thus the hash is mapped to the domain block one by one like the ER, if the is no gap between the head ER and the best domain block then the number of hash is also bounded by `BlockTreePruningDepth`.

Assume `BlockTreePruningDepth` is 10, the head ER is block #50 and the best domain block is #N, the hashes we need to maintain are `block #40, ..., #N` if there is no gap `N == 50` then there are only 10 hashes need to maintain.

Reply:
Yeah, I see, now that there is only one receipt per bundle, the gap can't be fixed once formed, this is another problem though. My point here is we don't have to track any consensus hashes in the state, they are redundant, and we can retrieve them from the client using a host function. We should be cautious of using the host function, but it's more straight here IMO.

Reply:
When using the host function we will have to handle the chain fork on the client side which is also tricky. If we can fix the gap problem then the hashes storage is bounded by `BlockTreePruningDepth` per domain but it is indeed redundant.

Reply:
Not necessarily, the runtime is tied to a specific block hash, so you don't have to deal with the forks.

Reply:
You can extend the input using some block hash stored in the System pallet to help identify the exact branch and find the correct hash corresponding to the number.

No need to have a perfect solution in this PR, we can add a TODO, leave it as is, and revisit it later as the issue of the gap needs to be handled anyway.

Reply:
> You can extend the input using some block hash stored in the System pallet to help identify the exact branch and find the correct hash corresponding to the number.

That is what I mean by tricky :P I will leave a TODO for now and open a new issue about the gap problem then we continue the discussion there.